File: algoritmosDeBusca/binarySearch.c
#include <stdio.h>

int binarySearch(int *v, int size, int element)
{
    int begin = 0;
    int end = size - 1;
    int middle;

    while (begin <= end)
    {
        middle = (end + begin) / 2;
        if (v[middle] < element)
        {
            begin = middle + 1;
        }
        if (v[middle] > element)
        {
            end = middle - 1;
        }
        if (v[middle] == element)
        {
            return middle;
        }
    }
    return -1;
}

int main()
{

    int aarray[10] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

    int saida = binarySearch(aarray, 10, 3);

    printf("%d", saida);
}

File: algoritmosOrdena/binarySearch.hs
import Data.Array

binarySearch :: Array Int Int -> Int -> Int -> Int -> Int 

binarySearch aarray element begin end
    | begin > end = -1
    | aarray ! middle < element = binarySearch aarray element (middle+1) end
    | aarray ! middle > element = binarySearch aarray element begin (middle-1)
    | otherwise = middle
    where
        middle = (begin + end) `div` 2

main :: IO ()
main = do

    let aarray = array (0,9) [(0, 1), (1, 2), (2, 3), (3, 4), (4, 5), (5, 6), (6, 7), (7, 8), (8, 9), (9, 10)]
    let result = binarySearch aarray 3 0 9

    print result